crates/core/asset/src/balance.rs

```diff
@@ -29,15 +28,154 @@ use decaf377::{r1cs::ElementVar, Fq, Fr};
+                }
+                // Subsequent entries for this asset ID in BTreeMap
+                //
+                //  1. Asset ID has a Required imbalance - accumulate another required amount
+                //  2. Asset ID has a Required imbalance - accumulate another provided amount
+                //  3. Asset ID has a Provided imbalance - accumulate another required amount
+                //  4. Asset ID has a Provided imbalance - accumulate another provided amount
+                std::collections::btree_map::Entry::Occupied(mut entry) => {
+                    let existing = entry.get_mut();
+                    match (existing, signed_value.negated) {
+                        (Imbalance::Required(existing_amount), true) => {
+                            *existing_amount = NonZeroU128::new(
+                                existing_amount
+                                    .get()
+                                    .checked_add(amount.get())
+                                    .ok_or_else(|| anyhow::anyhow!("Combining required amounts"))?,
+                            )
+                            .ok_or_else(|| anyhow::anyhow!("Combining required amounts"))?;
+                        }
+                        (Imbalance::Required(existing_amount), false) => {
+                            match existing_amount.get().checked_sub(amount.get()) {
+                                Some(diff) if diff > 0 => {
+                                    *existing_amount = NonZeroU128::new(diff)
+                                        .ok_or_else(|| anyhow::anyhow!("Reduce required amount"))?;
+                                }
+                                _ => {
+                                    *entry.get_mut() = Imbalance::Provided(
+                                        NonZeroU128::new(amount.get() - existing_amount.get())
+                                            .ok_or_else(|| {
+                                                anyhow::anyhow!(
+                                                    "Convert required amount to provided amount"
+                                                )
+                                            })?,
+                                    );
+                                }
+                            };
+                        }
+                        (Imbalance::Provided(existing_amount), true) => {
+                            match existing_amount.get().checked_sub(amount.get()) {
+                                Some(diff) if diff > 0 => {
+                                    *existing_amount = NonZeroU128::new(diff)
+                                        .ok_or_else(|| anyhow::anyhow!("Reduce provided amount"))?;
+                                }
+                                _ => {
+                                    *entry.get_mut() = Imbalance::Required(
+                                        NonZeroU128::new(amount.get() - existing_amount.get())
```

Comment:
Circling back to this later, I want to understand why this cannot be equal to zero (and thus, why it is infallible)

Reply:
I think it can be 0, you can send a proto with [-x UM, +x UM] and I think this method will panic?

Reply:
yup, the inverse case, [+x UM, -x UM] is also problematic. Explicitly handling the case where the difference is 0 and removing the entry from the map resolves this.